warmup.py

```diff
@@ -103,7 +105,23 @@ def energy(self, r):
             If ``r`` is 0, the energy is :py:obj:`numpy.inf`.
 
         """
-        pass
+        if isinstance(r, (list, tuple, np.ndarray)):
+            u = []
+            for ri in r:
```

Comment:
This implementation with a loop works, but python loops with appends are generally very slow. Can you try to find an implementation that only uses numpy features? In particular, try using selectors to pick out **all** instances where each of these cases occurs, and then evaluate once for each. You should probably consider using `np.zeros_like` to initialize `u`.